nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) 2020 Sarthak Mittal
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__BT_ACTION_SERVER_HPP_
+#define NAV2_BEHAVIOR_TREE__BT_ACTION_SERVER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <fstream>
+#include <set>
+#include <exception>
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_behavior_tree/behavior_tree_engine.hpp"
+#include "nav2_behavior_tree/ros_topic_logger.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+#include "nav2_util/simple_action_server.hpp"
+#include "rclcpp_action/rclcpp_action.hpp"
+#include "tf2_ros/buffer.h"
+#include "tf2_ros/transform_listener.h"
+#include "tf2_ros/create_timer_ros.h"
+
+namespace nav2_behavior_tree
+{
+/**
+ * @class nav2_behavior_tree::BtActionServer
+ * @brief An action server that uses behavior tree to execute an action
+ */
+template<class ActionT>
+class BtActionServer
+{
+public:
+  using ActionServer = nav2_util::SimpleActionServer<ActionT>;
+
+  typedef std::function<bool (typename ActionT::Goal::ConstSharedPtr)> OnGoalReceivedCallback;
+  typedef std::function<void ()> OnLoopCallback;
+
+  /**
+   * @brief A constructor for nav2_behavior_tree::BtActionServer class
+   */
+  explicit BtActionServer(
+    const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+    const std::string & action_name,
+    OnGoalReceivedCallback on_goal_received_callback,
+    OnLoopCallback on_loop_callback)
+  : action_name_(action_name),
+    node_(parent),
+    on_goal_received_callback_(on_goal_received_callback),
+    on_loop_callback_(on_loop_callback)
+  {
+    auto node = node_.lock();
+    logger_ = node->get_logger();
+    clock_ = node->get_clock();
+
+    const std::vector<std::string> plugin_libs = {
```

Comment:
This is just the default list for the `plugin_lib_names` parameter in case it's not provided by the user. Any BT autonomy system that uses this object can pass a list of specific plugins through the `plugin_lib_names` param. But the default list containing all the plugin names provided by nav2 should exist in the abstract server right?

Reply:
Would it be possible to have this be an input? Then each server can have its own default libs that it can give to the BT action server. So the parameter for plugin libs can be for the application and the value of it is given to the BT action server as an input to create its `BehaviorTreeEngine`

Reply:
This wasn't changed - what I was hoping for was the list of `libs` be an argument to the constructor of the BT action server client. That way the variable and the set of libs was under control of the BT action server's parent application. In this case, the application server is responsible for declare and getting them and just gives a `vector<string>` to it as an input so it doesn't worry about ROS variables.

Reply:
Got it!